Remove non-default defaults from mdadm resource

3dinfluence opened the pull request:
The purpose of this commit is to start a conversation about the current default values for the mdadm resource.  Which currently chooses some default values for optional parameters.  It seems odd for Chef to choose defaults for optional parameters where the underlaying OS/application has it's own defaults.

As a result this commit does not modify or extend the rspec tests to cover these changes.  I'll address this in a future commit pending the discussion.

This commit removes the default values for the following optional parameters when creating an mdadm device.

* chunk size: Chef 16k, mdadm 512k
* metadata version: Chef 0.90, mdadm 1.2

The metadata version has 3 notable impacts.

1. 0.90 is limited to 28 devices per RAID volume.
2. Max device size of 2TB.
3. Some older versions of grub and lilo only support RAID1 md devices with metadata 0.90 as a boot device.

3dinfluence commented:
@someara This is the PR that ties into one of the conversations we had on IRC the other day.  You asked me to ping you on it so here's the ping.  You also mentioned that @lamont-granquist should have a look at this.

lamont-granquist commented:
I like the idea, I'm a little concerned about breaking changes, you also seem to be have probably broken the specs that check these defaults are set.

We do probably need to at least announce via a warn that we're going to be changing to defaults for a bit first before changing them.

Are these default values that basically nobody should want these days (i.e. made sense back when this was written years ago and no longer make any sense at all?)

3dinfluence commented:
Yes I intentionally didn't address spec tests yet.  But the breaks are with tests that expect to see these defaults getting passed to the resource.  I wanted to get feedback on the idea before doing additional work to clean up the specs.

To be safe I think the warn route is a good idea. There's really no reason that this change needs to happen quickly.  It is possible that this change could break new systems if this resource is being used to create a boot volume and the version of grub on the system doesn't support the 1.2 metadata.  On existing systems  this change won't break any existing systems b/c the resource isn't 100% idempotent.  It only triggers the create action if the md device doesn't exist and there is no modify action.  Probably for good reason as making changes to an existing RAID could be destructive.

I'm sure these defaults made sense to someone or they wouldn't be in here to begin with.  It's  possible that metadata version 0.90 and 16k chunk size were the mdadm defaults at the time and it just leaked over into the Chef codebase.  The mdadm manpage certainly recommends using metadata 1.2 on all new installs.  And at this point there's probably more users overriding this default to get support for drives that are greater than 2TB and for building md devices with more than 28 devices than users who would have to set the metadata value back to 0.90 in order to support a boot device on older versions of grub.

Documentation would need to be updated as well.

lamont-granquist commented:
Yeah, I'm suspecting that while its technically breaking that this will solve more problems and cause substantially less waste of time than leaving the defaults where they are.  I'd suggest that if those attributes are not set on the mdadm resource that it warns that the default will be changing in the near future.  Should be real quick.  You should try to get that one in over the weekend if possible since we're going to try to release 12.1.0 next week.

3dinfluence commented:
Sounds good I'll work on that tonight once I'm finished shoveling snow.

3dinfluence commented:
I just tested this locally and I think I pushed it too quickly.  The warning is being shown even when a value is being passed to the resource for these attributes.

lamont-granquist commented:
Yeah when `new_resource.chunk` is called as an accessor in the provider you'll see that warning thrown.

You should move the warning into the provider.  Check for `new_resource.chunk.nil?` and warn there.  I'm not sure exactly where the best place to do that is -- define_resource_requirements is probably the most appropriate place.

lamont-granquist commented:
errr, wait, the default is set in the initializer of the resource, that makes it very annoying to know if the chunk of '16k' is the default value or is being set by the user.  

you'd really need to set up a new couple of accessors on the resource like 'user_set_chunk' which defaults to false, and then you set it to true if chunk is called with a non-nil value.  then in the provider in define_resource_requirements check `warn "..." unless user_set_chunk`

more evidence that i'd like to see default values on the resource stored separately from requested values on the resource so that `new_resource.chunk == "16k"` wasn't ambiguous about its source.

lamont-granquist commented:
So instead of the provider, you can use the after_created hook in the resource to do the test+warn.  This'll keep the validation logic better confined to the resource and not split between the resource + provider and so won't violate SRP as bad.

You'll still need some kind of flag to tell you if you the user set it or not.

3dinfluence commented:
Thanks I'll work on that.

3dinfluence commented:
This is now working correctly.  However I still have a question about testing these changes.

This works just fine without the def user_set_* methods, however, I added them back while trying to write some rspec tests to cover these changes.  I do have a couple of tests that pass that I haven't committed.  But it's just testing the default value of false for the user_set_* members.  Due to the nature of mocking up the resource I'm not sure how to mock up a meaningful test for a user passing a value into the resource.  Any suggestions?

lamont-granquist commented on lib/chef/resource/mdadm.rb:
```diff
@@ -39,11 +39,17 @@ def initialize(name, run_context=nil)
       end
 
       def chunk(arg=nil)
+        @user_set_chunk = true
```
that should be protected by `unless arg.nil?`

lamont-granquist commented on lib/chef/resource/mdadm.rb:
```diff
@@ -76,6 +82,7 @@ def level(arg=nil)
       end
 
       def metadata(arg=nil)
+        @user_set_metadata = true
```
ditto

lamont-granquist commented on lib/chef/resource/mdadm.rb:
```diff
@@ -99,6 +106,37 @@ def raid_device(arg=nil)
+          arg,
+          :kind_of => [ TrueClass, FalseClass ]
+        )
+      end
```
don't need these accessors, the flags are private state.

lamont-granquist commented on lib/chef/resource/mdadm.rb:
```diff
@@ -99,6 +106,37 @@ def raid_device(arg=nil)
+          'choose the default metadata version.'
+        chunk_warn_msg = 'The default chunk size of 16k will be removed in a '\
+          'future release.  This change will allow mdadm to choose '\
+          'the default chunk size.'
```
you might mention that the user can suppress the message by specifying the chunk size/metadata on their resource.

"#{self} the default chunk size of 16k will be removed in a future release.  To maintain backwards compatibility please explicitly specify the chunk size that you desire on the the mdadm resource."

Added a "#{self}" there as well so people know which resource is emitting the warn.

3dinfluence replied:
Do you think that wording may scare people?  This resource doesn't modify any existing raid devices.  So the planned change will only affect new devices created with the resource.  I don't want it to make anyone think that they have to make this change or risk breaking their existing md devices.

What do you think about this wording?
"#{self} default chunk size of 16k will be removed in a future release. To maintain backwards compatibility please explicitly specify the chunk size that you desire on the the mdadm resource if the mdadm default is not desired. This future change will only impact newly created md devices."

lamont-granquist replied:
Cool, yep, wordsmith it how you like.  I was mostly concerned that it not only emit the warning but make sure that the reader knows what action to take next.

nathenharvey commented:
@lamont-granquist can you assist with some advice on testing?

lamont-granquist commented:
The tests should go into spec/unit/resource/mdadm.rb and should create a resource and then setup an exepection that Chef::Log gets called with :warn.   spec/unit/resource/chef_gem.rb has some examples which test after_created hooks if the mdadm.rb resource is either not test well or doesn't have appropriate examples to copy.

3dinfluence commented:
@lamont-granquist thanks that got me going in the right direction.  The Chef::Log bit was what had me puzzled.  However I'm not having much luck putting all the pieces together.  I put a binding.pry in the resource's after_created method.  So I know that this code is getting called and the resource is in the expected state.  But my test is still failing.  Here's a gist showing what I have.  https://gist.github.com/3dinfluence/668bae56ddfaa27ca1c2

lamont-granquist commented:
Your expectation needs to come first, and get setup, then you need to call after_created.

3dinfluence commented:
@lamont-granquist Thanks that did the trick.  I've resync'ed my fork and pushed with my changes.  Thank you for all your help with his.

@nathenharvey I'm still waiting on our legal department at work for the CCLA.  The person who I'm working with on that is out of the office this week so I expect to have that squared away next week.

chef-supermarket commented:
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.chef.io/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## Non-GitHub Verified Committers

There are 1 commit author(s) whose commits are authored by a non-GitHub verified email address. Chef will have to manually verify that they are authorized to contribute.

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

3dinfluence commented:
Ok I should be covered under Dyn's CCLA now.

3dinfluence commented:
@lamont-granquist Do I need to do anything to get this patch moving forward again?  The CCLA should be all set now. https://supermarket.chef.io/ccla-signatures/275/contributors

lamont-granquist commented:
@chef/client-core needs additional review.

danielsdeleo commented:
:+1: feature-wise, it'd be cool if the user could opt-in to using the system defaults.

tyler-ball commented:
:+1: If we're going to change the defaults in a later release, it would be cool to add a test that reminds us to do that.

```ruby
it "has a default metadata of 1.2", :chef_13 => true do
  ...
end
```

btm commented:
@3dinfluence I see your github account is associated with a valid CCLA. Now make sure your dyn email address, which is on the commits, is added as an email address to your github account so the Curry bot can associate them.

Also, I agree with @tyler-ball, we always forget to find all those deprecations when we ship a major. It looks like we already have a ```:chef_gte_13_only``` rspec filter.

3dinfluence commented:
@btm I've added my Dyn email to my github account and made it the public email.  Not sure if that last step was necessary.  Once Curry bot has it associated can I safely change the public one back to my personal email address?  Looking at my git log for my commits my dyn.com email address is already associated with the commits.  Unless github does something to filter that when creating the PR from my fork.

I should probably update the title of this PR to make it in line with what this PR is actually doing.  This PR started out changing behaviour but based on feedback from @lamont-granquist what it actually does is just displays some messages to warn users of a future change to the default behaviour of the 2 optional parameters that this resource exposes: metadata version and chunk size.  Let me know if you would like me to that to get everything back on the same page here.

@danielsdeleo idea of adding an attribute to toggle this behaviour is interesting.  A new attribute would allow for this new behaviour to be tested now and allow old behaviour once the defaults are changed in the next major release.

@tyler-ball if my idea here is implemented then the new default for all optional parameters, which includes metadata version, will actually be nil.  But I get what you're saying.  I plan on submitting another PR that would implement the required changes rather than just display the warning messages that this PR does.  Should I add the appropriate rspec tests with the `:chef_gte_13_only` filter now with comments around the tests that reference the PR that will pull in the real changes so that when the time comes this doesn't fall through the cracks?

chef-supermarket commented:
Hi. Your friendly Curry bot here. Just letting you know that all commit authors have become authorized to contribute. I have added the "Signed CLA" label to this issue so it can easily be found in the future.

btm commented:
@3dinfluence I don't think it needs to be the email shown on your profile. I push from multiple emails depending on how old a computer is and curry seems fine with that provided they're all in my github. I guess non-public emails on github are still accessible through the API or something.

Anyway, I poked Curry manually to retry and it's happy now with this PR. Next time you push you should see the "Signed CLA" label show up automatically (you could change your primary email or whatever before your next push and answer your question).

thommay commented:
OK, this looks ready for merge. When it is merged, make sure that we add tests so that we deprecate properly in Chef 13 per btm above.

3dinfluence commented:
I'm not sure that I agree with the Ready for Merge label here.  I think there's still some unaddressed concerns that were raised.  Life got busy so I haven't had the time to circle back to this.  But summer is winding down and I should have more time.

I think it's still unclear if this is a change that is desired.  @danielsdeleo suggested adding a new attribute that toggles the behaviour between the old and this proposed behaviour.  I think what may be more widely excepted.

I think I like this approach myself after sitting on this issue for a while.  It allows this new behaviour to go out in the next minor version of chef so that people can opt into using it.  Giving some time to test the new behaviour.  If the default behaviour is going to change with Chef 13 then we can keep the warning log messages around this.  It also has the added benefit of simplifying this change if/when the time comes to change the default behaviour to simply changing the default value for this attribute.

Thoughts?

danielsdeleo commented:
As it is now, compatibility is maintained, but you don't get the thing you wanted (until 13.0), which is to be able to just use the system defaults. It'd be nice for users to be able to move to the new behavior now, if they desire it.

danielsdeleo commented:
@3dinfluence if you get time to clean this up, please at-mention me so I know to re-review.

lamont-granquist commented:
@3dinfluence okay taking off ready to merge, re-ping when you're ready again.

lamont-granquist commented:
paging @3dinfluence again?

3dinfluence commented:
@lamont-granquist I'm still about.  Sorry that I haven't gotten back to this.  I'll see what I can do in the next few days to get this back on track.

3dinfluence commented:
@lamont-granquist I rebased my changes on 12.7.2 but since there appears to be a conflict on master I can rebase it on master if that makes it easier.

These changes allow a user opt-out of the chef provided defaults but the current way this is implemented means that if the default value of mdadm_defaults is set to true in the future some care needs to make sure that the after_created function does the correct thing to enforce the old defaults.  I'm happy to rework this if you think we should try to make this easily changed.

It also doesn't allow the user to set chunk or metadata to nil without also setting mdadm_defaults to true.  That may be ok.

Let me know if you would like me to address either of these.  I'm not sure what appveyor is failing as I don't have an appveyor account and the message is getting cutoff in Github.

3dinfluence commented:
Forgot to @danielsdeleo to review this on my last comment.

danielsdeleo commented on lib/chef/resource/mdadm.rb:
```diff
@@ -98,6 +114,45 @@ def raid_device(arg = nil)
+          "will only impact newly created md devices."
+
+        Chef::Log.warn(metadata_warn_msg) unless @user_set_metadata
+        Chef::Log.warn(chunk_warn_msg) unless @user_set_chunk
```
This could be changed to `Chef::Log.deprecation`

danielsdeleo commented:
We should rebase on master. Other than, that, minor nitpick about the using the new deprecation logging stuff, but looks good.

3dinfluence commented:
@danielsdeleo It now uses the deprecation logging level.  Can you make sure that the spec tests cover that correctly b/c I had to add some additional expect statements to catch the deprecation errors.  Hopefully I did that correctly.  I also added a chef_gte_13_only check so that the resource gets flagged for the change in the default value of the mdadm_defaults attribute.

The rebase got my branch all out of sync so forced pushed it.  Sorry if that makes the reviewed changes more difficult to track.

danielsdeleo commented:
Force-pushing is normal after a rebase and is totally okay for feature branches (as long as you're not sharing the feature branch with other devs who have a copy of it on their machines). I'll look into the other questions in a sec.

danielsdeleo commented:
The way you handled `Chef::Log.deprecation` is fine. We have a config option set to turn deprecation warnings into errors in the tests to ensure we are not using deprecated functionality from one place somewhere else in core chef. The other way to handle that is to turn off the deprecation warnings as errors option just for the tests that need it.

danielsdeleo commented:
And finally, the chef_gte_13 test is fine. It will catch that there is a Chef 13.0 change to be made, which is all we need from it.

danielsdeleo commented:
Looks ready to merge to me. :+1: @chef/client-core anyone want to give this a final once-over?

coderanger commented on lib/chef/resource/mdadm.rb:
```diff
@@ -98,6 +114,49 @@ def raid_device(arg = nil)
+          "resource if the mdadm default is not desired. This future change "\
+          "will only impact newly created md devices."
+
+        Chef::Log.deprecation(metadata_warn_msg) unless @user_set_metadata
```
Should be `Chef.log_deprecation` I think.

coderanger commented:
The deprecation API is a bit wonky because we try to route them via the events system instead of direct logging if possible.

thommay commented:
needs tests fixing. :+1: from me once that's done though.

3dinfluence commented:
@thommay yeah looks like when I rebased it on master again the `:chef_gte_13_only` test started to run.  This test is supposed to fail as a reminder to change the default value for the mdadm_default attribute from false to true.  Did something recently change around how the `:chef_gte_13_only` tests are treated?

lamont-granquist commented:
use 'describe "in Chef-13", chef: ">= 13" do' now

3dinfluence commented:
@lamont-granquist I believe this is all set from my side unless anything else comes up.  Sorry for the delay in raising this back up.  Was travelling last week.

3dinfluence commented:
@lamont-granquist any update on this from your end?  I see that the branch has conflicts again.  If I resolve them can we get this merged?

3dinfluence commented:
After talks with @thommay at Chefconf we're going to close this PR and look at just adding an options attribute to the resource.
